File: driver/chrome.py
from os import makedirs
from subprocess import check_output
from brotli import decompress
from os.path import abspath, dirname, exists
from uuid import uuid4

import pyuseragents
from selenium import webdriver

_CHROMIUM_PATH = abspath(dirname(dirname(__file__))) + "/bin/chromium.br"
CHROMIUM_PATH = abspath(dirname(dirname(__file__))) + "/bin/chromium"
print("Chromium path:", CHROMIUM_PATH)
if not exists(CHROMIUM_PATH):
    with open(_CHROMIUM_PATH, "rb") as comp_chrome_file:
        with open(CHROMIUM_PATH, "wb") as chrome_file:
            chrome_file.write(decompress(comp_chrome_file.read()))
#CHROMIUM_PATH = check_output(["node", "driver/path.js"]).decode("utf-8").replace("\n", "")
#with open("./bin/path") as path_file:
#    CHROMIUM_PATH = path_file.read()
#CHROMIUM_PATH = "/tmp/chromium"
CHROMEDRIVER_PATH = abspath(dirname(dirname(__file__))) + "/bin/chromedriver"


class Chrome():
    def __init__(self, width: int = 1920, height: int = 1080) -> None:
        print("Initializing Chrome Object")
        self._width = int(width)
        self._height = int(height)
        self.chromium = str(CHROMIUM_PATH)
        self.chromedriver = str(CHROMEDRIVER_PATH)

        print("Creating temp folders")
        self._tmp_folder = '/tmp/{}'.format(uuid4())

        for folder in ["", "/user-data", "/data-path", "/cache-dir"]:
            if not exists(self._tmp_folder + str(folder)):
                makedirs(self._tmp_folder + str(folder))

        print("Defining the options")
        self.options = webdriver.ChromeOptions()
        for option in [
            '--headless',
            '--no-sandbox',
            '--disable-gpu',
            '--user-data-dir={}'.format(self._tmp_folder + '/user-data'),
            '--enable-logging',
            '--log-level=0',
            '--v=99',
            '--single-process',
            '--hide-scrollbars',
            '--data-path={}'.format(self._tmp_folder + '/data-path'),
            '--ignore-certificate-errors',
            '--homedir={}'.format(self._tmp_folder),
            '--disk-cache-dir={}'.format(self._tmp_folder + '/cache-dir'),
            'user-agent={}'.format(pyuseragents.random())
        ]:
            self.options.add_argument(option)

        print("Setting the Chromium Binary Location")
        self.options.binary_location = self.chromium

        print("Launching the driver")
        self.driver = webdriver.Chrome(options=self.options, executable_path=self.chromedriver)

    @property
    def width(self):
        print("Getting the width")
        _window_size = self.driver.get_window_size()
        print("Window size:", _window_size)
        self._width = int(_window_size.get("width", -1))
        self._height = int(_window_size.get("height", -1))
        return self._width

    @width.setter
    def width(self, width: int):
        if width is None:
            return
        print("Setting the width")
        self.driver.set_window_size(width, self._height)
        self._width = int(width)

    @property
    def height(self):
        print("Getting the height")
        _window_size = self.driver.get_window_size()
        print("Window size:", _window_size)
        self._width = int(_window_size.get("width", -1))
        self._height = int(_window_size.get("height", -1))
        return self._height

    @height.setter
    def height(self, height: int):
        if height is None:
            return
        print("Setting the height")
        self.driver.set_window_size(self._width, height)
        self._height = int(height)

    def open(self, url: str = None):
        if url is None:
            return
        print("Opening a new URL")
        self.driver.get(str(url))

    def close(self):
        print("Closing the driver")
        self.driver.close()

    def quit(self):
        print("Quitting the driver")
        self.driver.quit()

    def screenshot(self, url: str = None, width: int = None, height: int = None, base64: bool = False):
        self.open(url)
        self.width = width
        self.height = height
        if base64:
            print("Screenshotting as base64")
            return self.driver.get_screenshot_as_base64()
        else:
            print("Screenshotting as PNG")
            return self.driver.get_screenshot_as_png()

    def fullpage_screenshot(self, url: str = None, width: int = None, base64: bool = False):
        self.width = width
        self.open(url)
        _height = self.height
        self.height = self.get_page_height(url)
        if base64:
            print("Full page screenshotting as base64")
            screenshot = self.driver.get_screenshot_as_base64()
        else:
            print("Full page screenshotting as PNG")
            screenshot = self.driver.get_screenshot_as_png()
        self.height = _height
        return screenshot
    
    def source(self, url: str = None):
        self.open(url)
        print("Getting the page source code")
        return self.driver.page_source

    def get_page_height(self, url: str = None):
        self.open(url)
        print("Getting the page height")
        return int(self.driver.execute_script("return Math.max( document.body.scrollHeight, document.body.offsetHeight, document.documentElement.clientHeight, document.documentElement.scrollHeight, document.documentElement.offsetHeight )"))

File: driver/chrome.ts
import chromium from 'chrome-aws-lambda';
import log from "../utils/log";
import fs from "fs"
import { v4 as uuidv4 } from 'uuid';
import randomUserAgent from '../utils/useragents';
import { Browser, Page } from 'puppeteer-core';

export class Size {
    width: number
    height: number
    constructor(width: number = null, height: number = null) {
        this.width = width
        this.height = height
    }
}

export class Chrome {
    _init: boolean = false
    _size: Size
    _tmpFolder = "/tmp/" + uuidv4()
    options = chromium.args
    driver: Browser;
    page: Page;
    async init(width: number = 1920, height: number = 1080) {
        /* Chrome constructor */
        if (this._init) {
            return
        }
        log("Initializing Chrome Object", "Driver")
        if (!width) {
            width = 1920
        }
        if (!height) {
            height = 1080
        }
        this._size = new Size(width = width, height = height)
        log("Creating temp folders", "Driver")
        for (var folder in ["", "/user-data", "/data-path", "/cache-dir"]) {
            if (!fs.existsSync(this._tmpFolder + folder)) {
                fs.mkdirSync(this._tmpFolder + folder)
            }
        }

        log("Defining the options", "Driver")
        this.options.concat([
            "--user-data-dir=" + this._tmpFolder + "/user-data",
            "--data-path=" + this._tmpFolder + "/data-path",
            "--disk-cache-dir=" + this._tmpFolder + "/cache-dir",
            "--homedir=" + this._tmpFolder,
            "--user-agent=" + randomUserAgent(),
            "--window-size=" + this._size.width.toString() + "," + this._size.height.toString()
        ])

        log("Running chromium", "Driver")
        this.driver = await chromium.puppeteer.launch({
            args: this.options,
            defaultViewport: chromium.defaultViewport,
            executablePath: await chromium.executablePath,
            //executablePath: "/Applications/Google Chrome Dev.app/Contents/MacOS/Google Chrome Dev",
            headless: chromium.headless,
            ignoreHTTPSErrors: true,
        });
        this.page = await this.driver.newPage()
        await this.setSize(this._size)
        this.page.setCookie({name: "CONSENT", value: "YES+cb", url: "https://google.com"})
        this._init = true
    }

    get size() {
        log("Getting the window size", "Driver")
        let _viewport = this.page.viewport()
        this._size = new Size(_viewport.width, _viewport.height)
        return this._size
    }
    async setSize(size: Size) {
        if (size) {
            log("Setting the window size", "Driver")
            if (!size.width) {
                size.width = this._size.width
            }
            if (!size.height) {
                size.height = this._size.height
            }
            await this.page.setViewport({ width: size.width, height: size.height })
            this._size = size
        }
    }
    set size(size: Size) {
        this.setSize(size)
    }

    async open(url: string = null) {
        if (url) {
            log("Opening a new URL", "Driver")
            await this.page.goto(url, {
                waitUntil: "networkidle0",
            });
        }
    }

    async close() {
        await this.driver.close();
    }

    async screenshot(url: string = null, base64: boolean = false, full: boolean = false) {
        await this.open(url)
        if (base64) {
            log("Screenshotting as base64", "Driver")
            return await this.page.screenshot({ fullPage: full, encoding: "base64" })
        } else {
            log("Screenshotting as PNG", "Driver")
            return await this.page.screenshot({ fullPage: full, encoding: "binary" })
        }
    }

    async source(url: string = null) {
        await this.open(url)
        log("Getting the page source code")
        return await this.page.content()
    }
}
